### PR TITLE
WDPD-215 Handle address change

### PR DESCRIPTION
**Pull request by ps-cgrach**
### This PR

Handle address change correctly
Add checkbox for saving credit card data

### Jira Ticket
https://jira.parkside.at/browse/WDPD-215

### CODE REVIEW

File: metadata.php [hunk 1]
```diff
@@ -231,7 +231,6 @@
         'live_chat.tpl' => 'wirecard/paymentgateway/views/admin/inc/live_chat.tpl',
         'sepa_mandate_modal.tpl' => 'wirecard/paymentgateway/views/blocks/sepa_mandate_modal.tpl',
         'sepa_mandate.tpl' => 'wirecard/paymentgateway/views/blocks/sepa_mandate.tpl',
-        'tab_descriptor.tpl' => 'wirecard/paymentgateway/views/admin/tpl/tab_descriptor.tpl',
```
Comment (ps-holler) [inline metadata.php]: Why was this removed?

File: views/blocks/wirecard_credit_card_fields.tpl [hunk 2]
```diff
@@ -31,14 +32,24 @@
...
-    <input type="hidden" name="ord_agb" value="0">
+        <input type="hidden" name="ord_agb" value="0">
       [{/if}]
       <input type="hidden" name="oxdownloadableproductsagreement" value="0">
       <input type="hidden" name="oxserviceproductsagreement" value="0">
       <div id="creditcard-form-div"></div>
+
+      [{if $user->hasAccount()}]
+        <div style="margin-bottom: 10px">
+          <input type="checkbox" name="save_checkbox" id="save-checkbox"/>
```
Comment (ps-holler) [inline views/blocks/wirecard_credit_card_fields.tpl]: If you put the input element inside the label, you can omit the `id` attribute. This might be useful to avoid ID conflicts (`save-checkbox` could already be taken).

```html
<label>
  <input type="checkbox" name="save_checkbox"/> [{oxmultilang ident="wd_vault_save_text"}]
</label>
```

File: metadata.php (hunk 1 above)
Comment (ps-amaureder) [reply metadata.php]: because it was added twice

File: default_payment_config.xml [hunk 3]
```diff
@@ -36,42 +36,42 @@
...
-      <terms>1</terms>
-      <allowed_currencies>
-        <currency>EUR</currency>
-      </allowed_currencies>
-      <httpuser_eur>16390-testing</httpuser_eur>
-      <httppass_eur>3!3013=D3fD8X7</httppass_eur>
-      <maid_eur>2048677d-57f4-44b0-8d67-9014c6631d5f</maid_eur>
-      <secret_eur>74bd2f0c-6d1b-4e9a-b278-abc34b83ab9f</secret_eur>
-      <initial_title>wd_heading_title_payolution_b2b</initial_title>
+        <oxid>wdpayolution-b2b</oxid>
```
Comment (ps-dfritzsch) [inline default_payment_config.xml]: @tomazpu Double-checked that only the indentation changed, no logic.

File: Core/Vault.php [hunk 4]
```diff
@@ -30,28 +29,29 @@ class Vault
...
+        $aCards = self::_getCardsFromDb($sUserId, self::_getAddressId($oOrder));
 
         return array_filter($aCards, function ($aCard) {
             $oDateExpiration = new DateTime($aCard['EXPIRATIONYEAR'] . '-' . $aCard['EXPIRATIONMONTH'] . '-01');
-            $oDateExpiration->add(new DateInterval('P6M'));
-
             $oDateToday = new DateTime();
 
-            return $oDateToday < $oDateExpiration;
+            $bIsValid = $oDateToday < $oDateExpiration;
```
[the remark this answers is not in the document]
Comment (ps-amaureder) [reply Core/Vault.php]: < is correct but $oDateExpiration should be end of the shown month, i will change it

File: Core/OrderHelper.php [hunk 5]
```diff
@@ -401,18 +416,45 @@ public static function onBeforeOrderCreation($oPayment)
...
-        $sOrderId = self::_getLastOrderIdFromDb($sUserId);
         $oOrder = oxNew(Order::class);
-        $oOrder->load($sOrderId);
-        return $oOrder->getDelAddressInfo();
+
+        if (!$oOrder->load(self::_getLastOrderIdFromDb($sUserId))) {
+            return null;
+        };
+
+        if (!empty($oOrder->oxorder__oxdelcountryid->value)) {
```
Comment (tomazpu) [inline Core/OrderHelper.php]: The two returns are identical except one parameter. Please refactor it, one solution is making a variable of what is returned and doing it with array merge.

File: Core/OrderHelper.php [hunk 6]
```diff
@@ -429,4 +471,44 @@ private static function _getLastOrderIdFromDb($sUserId)
...
+     * @return array
+     *
+     * @since 1.3.0
+     */
+    public static function getSelectedShippingAddress()
+    {
+        $oOrder = oxNew(Order::class);
+        $oCurrentAddress =  $oOrder->getDelAddressInfo();
+
+        if (!is_null($oCurrentAddress)) {
```
Comment (tomazpu) [inline Core/OrderHelper.php]: Also this is not very clean. Again the same parameters are returned  just the parameter is named differently. You can split it up into a separate method where the data is filled up you pass into it the $oCurrentAddress or the $oUser depending on an if.

File: Core/Vault.php [hunk 7]
```diff
@@ -189,4 +192,19 @@ private static function _getAddressId($oOrder)
...
+     * @param array $aCard
+     * @param array $aVaultCard
+     *
+     * @return bool
+     *
+     * @since 1.3.0
+     */
+    private static function _areCardsEqual($aCard, $aVaultCard)
+    {
+        return $aCard['TOKEN'] === $aVaultCard['token']
```
Comment (tomazpu) [inline Core/Vault.php]: why are here the array fields once in uppercase the other time in camel case?

File: default_payment_config.xml (hunk 3 above)
Comment (tomazpu) [reply default_payment_config.xml]: @ps-dfritzsch Thank you

File: views/blocks/wirecard_credit_card_fields.tpl [hunk 8]
```diff
@@ -31,14 +32,24 @@
       <input type="hidden" name="challenge" value="[{$challenge}]">
       <input type="hidden" name="sDeliveryAddressMD5" value="[{$oView->getDeliveryAddressMD5()}]">
       [{if $oView->isActive('PsLogin') || !$oView->isConfirmAGBActive()}]
-    <input type="hidden" name="ord_agb" value="1">
+        <input type="hidden" name="ord_agb" value="1">
```
Comment (tomazpu) [inline views/blocks/wirecard_credit_card_fields.tpl]: is it not possible to put the if statement as a print to the value? then we have only one input in the code?

File: Controller/NotifyHandler.php [hunk 9]
```diff
@@ -292,4 +295,53 @@ private function _shouldUseTransactionID($sTransactionId, $sPaymentId)
...
+
+    /**
+     * @param SuccessResponse $oResponse
+     * @param Order           $oOrder
+     *
+     * @return bool
+     *
+     * @since 1.3.0
+     */
+    private static function _shouldSaveCard($oResponse, $oOrder)
```
Comment (tomazpu) [inline Controller/NotifyHandler.php]: Please add a method description on what you are checking for. Like Check if the payment method is credit card the user has an account and ...

File: Core/OrderHelper.php (hunk 5 above)
Comment (ps-amaureder) [reply Core/OrderHelper.php]: the first return, returns the delivery address, the second one the billing address

File: Core/OrderHelper.php (hunk 6 above)
Comment (ps-amaureder) [reply Core/OrderHelper.php]: here the first one is from the address db table, the second from the user db table

File: Core/Vault.php (hunk 7 above)
Comment (ps-amaureder) [reply Core/Vault.php]: $aCard is created by the DatabaseProvider and here the keys are uppercase, $aVaultCard is created by us and there we use our coding standard

File: Controller/NotifyHandler.php (hunk 9 above)
Comment (ps-amaureder) [reply Controller/NotifyHandler.php]: I added a description

File: views/blocks/wirecard_credit_card_fields.tpl (hunk 8 above)
Comment (ps-dfritzsch) [reply views/blocks/wirecard_credit_card_fields.tpl]: This code has not been changed (only its indentation), which is why we would consider it as an out-of-scope refactoring that should not be part of this PR.

File: Core/OrderHelper.php (hunk 5 above)
Comment (tomazpu) [reply Core/OrderHelper.php]: But that not a good solution, if one method decides on which data to build, then builds the "same" but two returns that are to 98% equal.

File: Core/OrderHelper.php (hunk 6 above)
Comment (tomazpu) [reply Core/OrderHelper.php]: Here i see the same issue as with the method above.

File: Core/OrderHelper.php (hunk 5 above)
Comment (ps-dfritzsch) [reply Core/OrderHelper.php]: Although these 2 returns look like they are 98% similar, they are not.
1. The first one uses OXID’s magic getters to build a shipping address using the **delivery** DB address data.
2. The second one uses OXID’s magic getters to build a shipping address using the **billing** DB address data

However, we would suggest to split it into **2 functions** to make the difference more obvious (`getDeliveryShippingAddressFromOrder` & `getBillingShippingAddressFromOrder`).

File: Core/OrderHelper.php (hunk 6 above)
Comment (ps-dfritzsch) [reply Core/OrderHelper.php]: see answer: https://github.com/wirecard/oxid-ee/pull/174/files/262fcf6670047123b8361756463fbd63c1e4c598#r307258277

However, we would also suggest to split it into **2 functions** to make the difference more obvious (`getCurrentDeliveryAddress` & `getUserAddress`).

File: Core/OrderHelper.php (hunk 5 above)
Comment (tomazpu) [reply Core/OrderHelper.php]: Yes please refactor.

$oOrder->oxorder__oxdelfname->value
$oOrder->oxorder__oxbillfname->value

difference is 'del' is replaced with 'bill', so the optimal solution is 3 methods (getDeliveryShippingAddressFromOrder & getBillingShippingAddressFromOrder and a generic one that gets the 'del/bill') to set the data.

Comment (ps-dfritzsch) [reply Core/OrderHelper.php]: Ok, we will split it up into **multiple methods** to make it more readable. 👍 

However, we do not think that creating a generic method for fetching 'del/bill' is a good practice as it would mean to avoid using **OXID's magic DB getters** (e.g. oxorder__oxdelfname), thus making it more complex and less readable/understandable.

File: Controller/NotifyHandler.php [hunk 10]
```diff
@@ -292,4 +295,57 @@ private function _shouldUseTransactionID($sTransactionId, $sPaymentId)
...
+        } catch (DatabaseConnectionException $oExc) {
+            $oLogger->error("Cannot save card", [$oExc]);
+        } catch (DatabaseErrorException $oExc) {
+            $oLogger->error("Cannot save card", [$oExc]);
+        }
+    }
+
+    /**
+     * Checks if the credit card credentials should be saved.
+     * For that the user needs an account, it has to be a credit card payment and the user must had checked the save
```
Comment (ps-dfritzsch) [inline Controller/NotifyHandler.php]: typo: must ~~had~~ **have** checked

Please also add a '.' at the end of the sentence.
